pkg/report/decompile.go

```diff
@@ -0,0 +1,150 @@
+const (
+	FlagForceArmThumbMode DecompilerFlag = "force_arm_thumb"
+)
+
+const objdumpCallTimeout = 10 * time.Second
+
+type DecompiledOpcode struct {
+	Offset  int
+	Len     int
+	Command string
+	IsBad   bool
+	// Yes, the field below duplicates information and can be inferred from the other fields,
+	// but at this point it does not seem to be reasonable to parse a string and then build
+	// exactly that string from its parts.
+	FullDescription string
+}
+
+// Decompiles a byte array with opcodes into human-readable descriptions.
+// Target must specify the environment from which the opcodes were taken.
+func DecompileOpcodes(raw []byte, flags []DecompilerFlag, target *targets.Target) ([]DecompiledOpcode, error) {
+	args, err := objdumpBuildArgs(flags, target)
+	if err != nil {
+		return nil, err
+	}
+
+	outLines, err := objdumpExecutor(raw, args, target)
+	if err != nil {
+		return nil, err
+	}
+
+	list := objdumpParseOutput(outLines)
+	if len(list) == 0 && len(raw) > 0 {
+		return nil, fmt.Errorf("no instructions found while the total size is %v bytes", len(raw))
+	}
+	return list, nil
+}
+
+func objdumpExecutor(raw []byte, args []string, target *targets.Target) ([]string, error) {
+	file, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp file: %v", err)
+	}
+	defer os.Remove(file.Name())
+	defer file.Close()
+
+	_, err = file.Write(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write to temp file: %v", err)
+	}
+
+	args = append(args, file.Name())
+	cmd := osutil.Command(target.Objdump, args...)
+
+	respBytes, err := osutil.Run(objdumpCallTimeout, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	return strings.Split(string(respBytes), "\n"), nil
+}
+
+// nolint: lll
+var (
+	objdumpAsmLineRegexp = regexp.MustCompile(`\s+([a-fA-F0-9]+)\:\s+((?:[a-fA-F0-9]{2,8}\s*)*[a-fA-F0-9]{2,8})\s+(.*?)\s*$`)
+	objdumpSpaceRegexp   = regexp.MustCompile(`\s`)
+)
+
+const objdumpBadCommand = "(bad)"
+
+func objdumpParseOutput(lines []string) []DecompiledOpcode {
+	ret := []DecompiledOpcode{}
+	for _, line := range lines {
+		result := objdumpAsmLineRegexp.FindStringSubmatch(line)
+		if result == nil {
+			continue
+		}
+		chunkLen := len(objdumpSpaceRegexp.ReplaceAllString(result[2], "")) / 2
+		offset, err := strconv.ParseUint(result[1], 16, 64)
+		if err != nil {
+			continue
+		}
+		ret = append(ret, DecompiledOpcode{
+			Offset:          int(offset),
+			Len:             chunkLen,
+			Command:         result[3],
+			IsBad:           result[3] == objdumpBadCommand,
+			FullDescription: result[0],
+		})
+	}
+	return ret
+}
+
+func objdumpBuildArgs(flags []DecompilerFlag, target *targets.Target) ([]string, error) {
+	hasFlag := func(flag DecompilerFlag) bool {
+		for _, val := range flags {
+			if val == flag {
+				return true
+			}
+		}
+		return false
+	}
+
+	ret := []string{"-b", "binary", "-D"}
+	switch target.Arch {
+	case targets.ARM64:
+		ret = append(ret, "-maarch64")
```

Comment:
Do we need the arch flag? In pkg/cover/backend/dwarf.go we don't add arch flag and it works. I think the objdump binary with target prefix already assumes the right arch.

Reply:
It seems this function can be reduced to:
```
args := []string{"-b", "binary", "-D"}
if arch == targets.ARM && flags&FlagForceArmThumbMode != 0 {
  args = append(args, "-M", "force-thumb")
}
```
and simply inlined into objdumpExecutor.

Reply:
It seems that it works for dwarf.go because in that case it receives ELF files, where architecture is already specified. 

Running `objdump -b binary -D <file>` gave me the `objdump: can't disassemble for architecture UNKNOWN!` error

Reply:
Ah, I see. Please add a comment that -march is mandatory.